File: server/src/readjson/StatTower.java
package readjson;

import config.battle.BattleConfig;

import java.io.Serializable;
import java.util.HashMap;

public class StatTower extends StatCard {
    private String archetype;
    private String targetType;
    private int auraTowerBuffType;
    private String bulletType;
    private int bulletTargetBuffType;
    private int attackAnimationTime;
    private int shootAnimationTime;
    private HashMap<Integer, StatEvolution> stat;

    public int getTargetType() {
        if (BattleConfig.TOWER_TARGET_TYPE.containsKey(this.targetType))
            return BattleConfig.TOWER_TARGET_TYPE.get(this.targetType);
        return BattleConfig.TOWER_TARGET_TYPE_NONE;
    }

    public int getAuraTowerBuffType() {
        return this.auraTowerBuffType;
    }

    public int getBulletTargetBuffType() {
        return this.bulletTargetBuffType;
    }

    public double getAttackAnimationTime() {
        return this.attackAnimationTime / 1000.0;
    }

    public double getShootAnimationTime() {
        return this.shootAnimationTime / 1000.0;
    }

    public double getDamage(int evolution, int level) {
        return this.stat.get(evolution).damage * (1 + 0.1 * level);
    }

    public double getRange(int evolution) {
        return this.stat.get(evolution).range;
    }

    public double getBulletRadius(int evolution) {
        return this.stat.get(evolution).bulletRadius;
    }

    public double getAttackSpeed(int evolution) {
        return this.stat.get(evolution).attackSpeed / 1000.0;
    }

    public double getBulletSpeed(int evolution) {
        return this.stat.get(evolution).bulletSpeed / 10;
    }
}

class StatEvolution implements Serializable {
    double damage;
    double range;
    double bulletRadius;
    double attackSpeed;
    double bulletSpeed;
}

File: client/src/modules/readjson/StatTower.js
/**
 * Created by Team 2 - LongLH - GDF 17 on 12/12/2022.
 */


var StatTower = StatCard.extend({

    archetype: String.prototype,
    targetType: String.prototype,
    auraTowerBuffType: Number.prototype,
    bulletType: String.prototype,
    bulletTargetBuffType: Number.prototype,
    attackAnimationTime: Number.prototype,
    shootAnimationTime: Number.prototype,
    stat: Map.prototype,

    getTargetType: function () {
        if (this.targetType in TOWER.TARGET_TYPE)
            return TOWER.TARGET_TYPE[this.targetType];
        return TOWER.TARGET_TYPE_NONE;
    },

    getAuraTowerBuffType: function () {
        return this.auraTowerBuffType;
    },

    getBulletTargetBuffType: function () {
        return this.bulletTargetBuffType;
    },

    getAttackAnimationTime: function () {
        return this.attackAnimationTime / 1000.0;
    },

    getShootAnimationTime: function () {
        return this.shootAnimationTime / 1000.0;
    },

    getDamage: function (evolution, level) {
        return this.stat[evolution].damage * (1 + 0.1 * level);
    },

    getRange: function (evolution) {
        return this.stat[evolution].range;
    },

    getBulletRadius: function (evolution) {
        return this.stat[evolution].bulletRadius;
    },

    getAttackSpeed: function (evolution) {
        return this.stat[evolution].attackSpeed / 1000.0;
    },

    getBulletSpeed: function (evolution) {
        return this.stat[evolution].bulletSpeed / 10;
    },

});

var StatEvolution = cc.Class.extend({
    damage: Number.prototype,
    range: Number.prototype,
    bulletRadius: Number.prototype,
    attackSpeed: Number.prototype,
    bulletSpeed: Number.prototype,
});